### ISSUE TITLE
problem with highly compressed data (memory access violation)

### ISSUE THREAD

**Issue by DavidBrenner3** [lz4/lz4#343]
Hi,

On some data with high "compressibility" (like 8 bit grayscale bitmaps),
_LZ4_decompress_fast throws (memory) access violation.
If I don't set the size of the array with the compressed data to the value returned by LZ4_compress/LZ4_compressHCVar or if I use LZ4_decompress_safe, seems to be ok.

I know _LZ4_decompress_fast it's not that safe to use, but, still, it shouldn't do that.

Could you please fix it?

Thank you.

Regards,
David

Btw, I tried to attach the bitmap compressed as zip, but git wouldn't accepted i (even if it says it does support ZIP).
I uploaded it as png (please decompress it to 8 bit grayscale bitmap before testing).

![problem](https://cloud.githubusercontent.com/assets/9998134/24387598/aac5485e-137e-11e7-913f-7ce26622d3c4.png)

**Comment by Cyan4973** [comment]
`LZ4_decompress_fast()` prototype is : 
`int LZ4_decompress_fast (const char* source, char* dest, int originalSize);`

This API works under the following condition : 
- `originalSize` is the _exact size_ of object to decompress (in this case, an 8-bits grayscale image in bitmap format).
- `dest` buffer is at least as large as `originalSize`
- `source` buffer is at least as large as the compressed object, and contain the entire compressed object.
- the compressed object is "clean", no error, no modification to its content.

If any of above conditions is not respected, this prototype will fail, possible throwing memory access violation.  

If you feel all above conditions are correctly respected, and `LZ4_decompress_fast()` nonetheless fail, then indeed, there is a problem that should be fixed.

**Comment by DavidBrenner3** [comment]
I already knew this.
And yes, the conditions are correctly respected.

**Comment by DavidBrenner3** [comment]
Well, in my application I've solved the problem: I replaced all calls to LZ4_decompress_fast with LZ4_decompress_safe.
Since most of the time the application is (re)compressing to LZ4 and to jpeg, there is no noticeable speed drop.

If you need additional help from me debugging the problem just let me know.
Offtopic: I've got my own problems too: I'm trying to find a way to use a lookup table with SSSE3.

**Comment by Cyan4973** [comment]
I'll keep the issue around, to get back to it when I've got time.

`LZ4_decompress_safe()` is definitely the recommended method,
its safety guarantees are well worth the small speed drop.

**Comment by Cyan4973** [comment]
I've been testing your image with `LZ4_decompress_fast()` lately.

I couldn't make it crash.
It's unclear to me why the result would be different.
I converted your tested image into greyscale bmp (size 2074746 bytes).
I tested compressing at different compression levels (ratio varied from 19.320 to 29.518), decompressing with `LZ4_decompress_fast()`, adding some sanitizer to ensure boundaries conditions, and even supervised execution with `valgrind`. It always worked, without a single memory access violation.

I'm unable to reproduce this bug.
